src/cryptonote_core/service_node_list.cpp

```diff
@@ -652,21 +659,58 @@ namespace service_nodes
   std::vector<crypto::public_key> service_node_list::get_expired_nodes(uint64_t block_height) const
   {
     std::vector<crypto::public_key> expired_nodes;
+    int hard_fork_version = m_blockchain.get_hard_fork_version(block_height);
 
-    const uint64_t lock_blocks = get_staking_requirement_lock_blocks(m_blockchain.nettype()) + STAKING_REQUIREMENT_LOCK_BLOCKS_EXCESS;
+    uint64_t lock_blocks = get_staking_requirement_lock_blocks(m_blockchain.nettype());
+    if (hard_fork_version >= cryptonote::Blockchain::version_10_swarms)
+      lock_blocks += STAKING_REQUIREMENT_LOCK_BLOCKS_EXCESS;
 
     if (block_height < lock_blocks)
       return expired_nodes;
 
-    for (auto &it : m_service_nodes_infos)
+    if (hard_fork_version >= cryptonote::Blockchain::version_10_swarms)
+    {
+      for (auto &it : m_service_nodes_infos)
+      {
+        crypto::public_key const &pubkey = it.first;
+        service_node_info  const &info   = it.second;
+
+        uint64_t node_expiry_height = info.registration_height + lock_blocks;
+        if (block_height >= node_expiry_height)
+        {
+          expired_nodes.push_back(pubkey);
+        }
+      }
+    }
+    else
     {
-      crypto::public_key const &pubkey = it.first;
-      service_node_info  const &info   = it.second;
+      const uint64_t expired_nodes_block_height = block_height - lock_blocks;
```

Comment:
Can't we use the `m_service_nodes_infos` in the `else` branch (as it is done above) to determine which nodes expired?

Reply:
They look equivalent but are not. There's something here that causes a consensus change, I've only tested briefly- so I haven't narrowed it down- the easiest way was to just guard the feature behind hardfork 10